### ISSUE TITLE
Could not resolve "virtual:nuxt-runtime-config"

### ISSUE THREAD

**Issue by Yakov-Chernogor** [nuxt-modules/storybook#718]
I'm getting this error on a clean nuxt project + storybook while running storybook in dev mode (npm run storybook)
Reproduce on stackblitz [https://stackblitz.com/edit/nuxt-storybook-error?file=package.json](https://stackblitz.com/edit/nuxt-storybook-error?file=package.json)
```
✘ [ERROR] Could not resolve "virtual:nuxt-runtime-config"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30:
      5 │ import { runtimeConfig } from 'virtual:nuxt-runtime-config';
        ╵                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "virtual:nuxt-runtime-config" as external to exclude it from the bundle,
  which will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/css"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7:
      6 │ import '#build/css';
        ╵        ~~~~~~~~~~~~

  You can mark the path "#build/css" as external to exclude it from the bundle, which will remove
  this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/plugins"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20:
      7 │ import plugins from '#build/plugins';
        ╵                     ~~~~~~~~~~~~~~~~

  You can mark the path "#build/plugins" as external to exclude it from the bundle, which will
  remove this error and leave the unresolved path in the bundle.


 ERROR  error while updating dependencies:                                                               9:10:45 PM
Error: Build failed with 3 errors:
node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30: ERROR: Could not resolve "virtual:nuxt-runtime-config"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7: ERROR: Could not resolve "#build/css"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20: ERROR: Could not resolve "#build/plugins"
    at failureErrorWithLog (/home/yakov/GitHub/KonTxT/KonTxT/node_modules/esbuild/lib/main.js:1472:15)
    at /home/yakov/GitHub/KonTxT/KonTxT/node_modules/esbuild/lib/main.js:945:25
    at /home/yakov/GitHub/KonTxT/KonTxT/node_modules/esbuild/lib/main.js:1353:9
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    ```

**Comment by tri-star** [comment]
@tobiasdiez
I have encountered the same issue as described in this Issue (I am using v8.1.5 and pnpm). The repository where I confirmed the reproduction with relatively small steps is as follows:
https://github.com/tri-star/storybook8-with-nuxtjs

After running `npx storybook-nuxt init`, I manually updated the versions of Storybook and other packages to the latest ones. 
I apologize if there were any mistakes in this process.

**Comment by obulat** [comment]
Yes, @tobiasdiez, I got the same errors again.

Since the latest changes weren't released yet, and I don't know how to install monorepo packages from the github main branch, I used `pnpm pack` to create a tar files for both packages and install them to Openverse Nuxt 3 branch locally. I got exactly the same errors as described in the first message of this issue in the console. The story names were listed in the sidebar, but the stories do not render.

The browser console shows the following error:
```
BW1NhrJ4.js:6 [nuxt] error caught during app initialization 
TypeError: Cannot read properties of undefined (reading 'cdnURL')
    at b (Ck1jFS18.js:1:4381)
    at et (Ck1jFS18.js:1:4345)
    at xv (BW1NhrJ4.js:2:19913)
    at Ws (BW1NhrJ4.js:2:20105)
    at uc (BW1NhrJ4.js:2:20140)
    at BW1NhrJ4.js:2:60933
    at n (BOf5JozU.js:1:6607)
    at Object.runWithContext (UqsEW9He.js:13:19498)
    at w (BOf5JozU.js:1:6645)
    at BOf5JozU.js:1:4665
    at Ys.run (UqsEW9He.js:9:267)
    at Object.runWithContext (BOf5JozU.js:1:4657)
    at BW1NhrJ4.js:2:60914
```
When navigating into the source files, it seems that the culprit is the `nitro.client.mjs` file with the following line:

https://github.com/nuxt/nuxt/blob/48a8b18083ecfc5b9514d6683a2958e7a1a41e13/packages/nuxt/src/core/templates.ts#L291

If `window` does not have a `__NUXT__` object here, then `useRuntimeConfig` returns an empty object `{}`. And then `paths.mjs` errors out when trying to access `useRuntimeConfig().app`:

https://github.com/nuxt/nuxt/blob/48a8b18083ecfc5b9514d6683a2958e7a1a41e13/packages/nuxt/src/core/templates.ts#L367-L372

**Comment by Blecus2k** [comment]
Can confirm, getting the same Errors..
Used the example/starter config and files...

`
 WARN  Sourcemap for "/virtual:/@storybook/builder-vite/setup-addons.js" points to missing source files                                                                                                                            08:01:02


 WARN  Sourcemap for "/virtual:/@storybook/builder-vite/vite-app.js" points to missing source files                                                                                                                                08:01:02

✘ [ERROR] Could not resolve "virtual:nuxt-runtime-config"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30:
      5 │ import { runtimeConfig } from 'virtual:nuxt-runtime-config';
        ╵                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "virtual:nuxt-runtime-config" as external to exclude it from the bundle,
  which will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/css"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7:
      6 │ import '#build/css';
        ╵        ~~~~~~~~~~~~

  You can mark the path "#build/css" as external to exclude it from the bundle, which will remove
  this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/plugins"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20:
      7 │ import plugins from '#build/plugins';
        ╵                     ~~~~~~~~~~~~~~~~

  You can mark the path "#build/plugins" as external to exclude it from the bundle, which will
  remove this error and leave the unresolved path in the bundle.


 ERROR  error while updating dependencies:                                                                                                                                                                                         08:01:03
Error: Build failed with 3 errors:
node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30: ERROR: Could not resolve "virtual:nuxt-runtime-config"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7: ERROR: Could not resolve "#build/css"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20: ERROR: Could not resolve "#build/plugins"

`

**Comment by agantelin** [comment]
I also get the same errors on a fresh Nuxt installation.

**Comment by yonglory-ms** [comment]
I'm also getting the same errors during the storybook setup.

**Comment by teonji** [comment]
I have the same problem. Is there any version or trick to avoid it for the time being so that I can start writing documentation?

**Comment by tian-zhihui** [comment]
After added the following code into .storybook/preview.ts, I can temporarily slove the problem in my project.
Hope my suggestions can help you
```
window.__NUXT__ = {
  serverRendered: false,
  config: {
    public: {},
    app: { baseURL: '/' },
  },
  data: {},
  state: {},
};
```

**Comment by thelfensdrfer** [comment]
> After added the following code into .storybook/preview.ts, I can temporarily slove the problem in my project. Hope my suggestions can help you

@tian-zhihui Sadly this does not work for me. I am using the default settings and the latest version `8.1.5`.

**Comment by teonji** [comment]
> > After added the following code into .storybook/preview.ts, I can temporarily slove the problem in my project. Hope my suggestions can help you
> 
> @tian-zhihui Sadly this does not work for me. I am using the default settings and the latest version `8.1.5`.

Unfortunately, it doesn't work for me either.

**Comment by tian-zhihui** [comment]
@thelfensdrfer @teonji 
Sorry I didn't explain the background clearly. 
I checked out the latest main branch, ran pnpm build and installed package "@nuxtjs/storybook" and "@storybook-vue/nuxt" locally.
Then I encountered the same problem like https://github.com/nuxt-modules/storybook/issues/718#issuecomment-2228817716.
I solved the `Cannot read properties of undefined (reading 'cdnURL')` problem using the above code.

**Comment by teonji** [comment]
> @thelfensdrfer @teonji Sorry I didn't explain the background clearly. I checked out the latest main branch, ran pnpm build and installed package "@nuxtjs/storybook" and "@storybook-vue/nuxt" locally. Then I encountered the same problem like [#718 (comment)](https://github.com/nuxt-modules/storybook/issues/718#issuecomment-2228817716). I solved the `Cannot read properties of undefined (reading 'cdnURL')` problem using the above code.

OK thanks for the explanation :) I get the error in the issue name and don't know how i can do it.
Are the two things related? is this not reproducible?

**Comment by melvinprindustry** [comment]
Running `npx nuxi@latest module add storybook` and afterwards `npm run dev` my startup time increased dramatically. Nuxt would start building itself and a big fat warning:
```
   A new version (8.2.4) is available!                               │
│                                                                     │
│   Upgrade now: npx storybook@latest upgrade
```
Would appear. So I ran `npx storybook@latest upgrade`. Startup time is good again, no weird building tricks. But now I get the exact same errors as OP:

```
✘ [ERROR] Could not resolve "virtual:nuxt-runtime-config"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30:
      5 │ import { runtimeConfig } from 'virtual:nuxt-runtime-config';
        ╵                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "virtual:nuxt-runtime-config" as external to exclude it from the bundle,
  which will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/css"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7:
      6 │ import '#build/css';
        ╵        ~~~~~~~~~~~~

  You can mark the path "#build/css" as external to exclude it from the bundle, which will remove
  this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/plugins"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20:
      7 │ import plugins from '#build/plugins';
        ╵                     ~~~~~~~~~~~~~~~~

  You can mark the path "#build/plugins" as external to exclude it from the bundle, which will
  remove this error and leave the unresolved path in the bundle.


 ERROR  error while updating dependencies:                                                                                                                                                         09:54:06
Error: Build failed with 3 errors:
node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30: ERROR: Could not resolve "virtual:nuxt-runtime-config"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7: ERROR: Could not resolve "#build/css"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20: ERROR: Could not resolve "#build/plugins"
    at failureErrorWithLog (/Users/melvinidema/development/portal-nuxt-3/portal/node_modules/esbuild/lib/main.js:1472:15)
    at /Users/melvinidema/development/portal-nuxt-3/portal/node_modules/esbuild/lib/main.js:945:25
    at /Users/melvinidema/development/portal-nuxt-3/portal/node_modules/esbuild/lib/main.js:1353:9
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    ```

The upgrade command upgraded the following packages:
```
- @nuxtjs/storybook: 7.0.2 => 8.1.5
- @storybook-vue/nuxt: 0.2.6 => 8.1.5
```

**Comment by tobiasdiez** [comment]
Is this specific to npm (and yarn), or does anyone experience it with pnpm?

**Comment by melvinprindustry** [comment]
> Is this specific to npm (and yarn), or does anyone experience it with pnpm?

Good question, we're using Yarn 4.1.1. I've tried getting the examples to work which require pnpm but with no luck so far.

**Comment by melvinprindustry** [comment]
# Steps to reproduce:

## Setup:
- MacOS Sonoma 14.3.1
- M1 Pro
- pnpm 9.5.0

Clone the git repository:
`git clone git@github.com:storybook-vue/nuxt-storybook-module-demo.git`

CD into the directory and run:
`pnpm install`

Then, run:
`pnpm dev`

The errors will appear.

**Comment by obulat** [comment]
> Is this specific to npm (and yarn), or does anyone experience it with pnpm?

I'm using pnpm

**Comment by JRew1234** [comment]
> Is this specific to npm (and yarn), or does anyone experience it with pnpm?

I do

**Comment by tobiasdiez** [comment]
Thanks everyone for your helpful responses!

I have a solution at https://github.com/nuxt-modules/storybook/pull/726. Still needs a bit of cleanup but it's working for me. Could you please try it out using the `pnpm pack` method described in https://github.com/nuxt-modules/storybook/issues/718#issuecomment-2228817716
I pretty busy the next few days but if this approach is working, I may find the time early next week to finish the PR.

**Comment by cheindl** [comment]
This fix seems to be a step forward as i don´t get the original errors from above ([ERROR] Could not resolve "virtual:nuxt-runtime-config").

But even in combination with the changes from #723 i get the following errors from the Preview iframe when opening a story:

```
[nuxt] error caught during app initialization
TypeError: Cannot read properties of undefined (reading 'cdnURL')
at publicAssetsURL (paths.mjs:8:32)
    at buildAssetsURL (paths.mjs:6:60)
    at fetchManifest (manifest.js?v=447340d6:12:21)
    at getAppManifest (manifest.js?v=447340d6:26:22)
    at getRouteRules (manifest.js?v=447340d6:35:9)
    at router.js?v=447340d6:155:65
    at fn (nuxt.js?v=447340d6:214:44)
    at Object.runWithContext (runtime-core.esm-bundler.js?v=447340d6:3944:18)
    at callWithNuxt (nuxt.js?v=447340d6:220:24)
    at nuxt.js?v=447340d6:37:41
```

Even if there is a window.__NUXT__ object present within the preview iframe, the configuration seems to be read to early within paths.mjs (`const appConfig = useRuntimeConfig().app` is outside the `publicAssetsURL` function).

**Comment by JRew1234** [comment]
@tobiasdiez works fine for me, I have some problems with story not rendering inside vuetify slot but I guess it's not related

**Comment by obulat** [comment]
With these changes, I get the following errors:
```
[Vue Router warn]: Record with path "/iframe.html" is either missing a "component(s)" or "children" property.
```
_comes from `applyPlugins` in `preview.mjs`_

```
[nuxt] error caught during app initialization TypeError: Cannot read properties of undefined (reading 'cdnURL')
    at publicAssetsURL (paths.mjs:8:32)
    at buildAssetsURL (paths.mjs:6:60)
    at fetchManifest (manifest.js:12:21)
    at getAppManifest (manifest.js:26:22)
    at getRouteRules (manifest.js:35:9)
    at router.js:161:42
    at fn (nuxt.js:214:44)
    at Object.runWithContext (runtime-core.esm-bundler.js:3104:18)
    at callWithNuxt (nuxt.js:220:24)
    at nuxt.js:37:41
```
_comes from `hookable/index.mjs`_

```
[Vue warn]: Component provided template option but runtime compilation is not supported in this build of Vue. Configure your bundler to alias "vue" to "vue/dist/vue.esm-bundler.js". 
  at <Anonymous> 
  at <Anonymous> 
  at <App>
```
_comes from `sb_preview/runtime.js`

**Comment by andresilva-cc** [comment]
I'm having the same problem while using Yarn Classic (v1).

**Comment by Nemhis** [comment]
Hi, I think I have the same problem when installing storybook 8.2.6

```
✘ [ERROR] Could not resolve "virtual:nuxt-runtime-config"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30:
      5 │ import { runtimeConfig } from 'virtual:nuxt-runtime-config';
        ╵                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "virtual:nuxt-runtime-config" as external to exclude it from the bundle,
  which will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/css"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7:
      6 │ import '#build/css';
        ╵        ~~~~~~~~~~~~

  You can mark the path "#build/css" as external to exclude it from the bundle, which will remove
  this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/plugins"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20:
      7 │ import plugins from '#build/plugins';
        ╵                     ~~~~~~~~~~~~~~~~

  You can mark the path "#build/plugins" as external to exclude it from the bundle, which will
  remove this error and leave the unresolved path in the bundle.


 ERROR  error while updating dependencies:                                                                                                                                                                                                                                                         11:17:31 PM
Error: Build failed with 3 errors:
node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30: ERROR: Could not resolve "virtual:nuxt-runtime-config"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7: ERROR: Could not resolve "#build/css"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20: ERROR: Could not resolve "#build/plugins"
    at failureErrorWithLog (/Users/pavelkurapov/projects/bikenet-front/node_modules/esbuild/lib/main.js:1472:15)
    at /Users/pavelkurapov/projects/bikenet-front/node_modules/esbuild/lib/main.js:945:25
    at /Users/pavelkurapov/projects/bikenet-front/node_modules/esbuild/lib/main.js:1353:9
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) (x2)
 ```
 
 There is my package.json
 ```json
 {
  "name": "nuxt-app",
  "private": true,
  "type": "module",
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix"
  },
  "dependencies": {
    "@healerlab/vue3-facebook-login": "^0.1.7",
    "@nuxt/eslint": "^0.3.13",
    "@nuxtjs/i18n": "^8.3.1",
    "@nuxtjs/storybook": "^8.1.5",
    "@pinia/nuxt": "^0.5.1",
    "@vee-validate/i18n": "^4.13.1",
    "@vee-validate/nuxt": "^4.13.1",
    "@vee-validate/rules": "^4.13.1",
    "@vueuse/core": "^10.11.0",
    "lodash-es": "^4.17.21",
    "nuxt": "^3.12.1",
    "nuxt-viewport": "^2.1.5",
    "nuxt-vue3-google-signin": "^0.0.11",
    "pinia": "^2.1.7",
    "vue": "^3.4.27",
    "vue-router": "^4.3.3",
    "vue3-recaptcha-v2": "2.0.2"
  },
  "devDependencies": {
    "@nuxtjs/svg-sprite": "1.0.2",
    "@types/lodash-es": "4.17.12",
    "@typescript-eslint/eslint-plugin": "^7.13.1",
    "@typescript-eslint/parser": "^7.13.1",
    "eslint": "^8.57.0",
    "sass": "^1.77.5",
    "sass-loader": "^14.2.1",
    "vite-svg-loader": "5.1.0"
  }
}
 ```
 
 Help pls, someone found workaround ?

**Comment by koya1616** [comment]
@Limonische 
In my environment

## working properly
```
    "@nuxtjs/storybook": "^8.1.1",
    "@storybook-vue/nuxt": "0.2.7",
    "@storybook/addon-essentials": "^8.2.4",
    "@storybook/addon-interactions": "^8.2.4",
    "@storybook/addon-links": "^8.2.4",
    "@storybook/blocks": "^8.2.4",
    "@storybook/builder-vite": "^8.2.4",
    "@storybook/test": "^8.2.4",
    "@storybook/vue3": "^8.2.4",
    "nuxt": "3.11.2",
```

## does not work
```
    "@nuxtjs/storybook": "^8.1.5",
    "@storybook-vue/nuxt": "^8.1.5",
    "@storybook/addon-essentials": "^8.2.6",
    "@storybook/addon-interactions": "^8.2.6",
    "@storybook/addon-links": "^8.2.6",
    "@storybook/blocks": "^8.2.6",
    "@storybook/builder-vite": "^8.2.6",
    "@storybook/test": "^8.2.6",
    "@storybook/vue3": "^8.2.6",
    "nuxt": "^3.12.4",
```

**Comment by andresilva-cc** [comment]
> @storybook/test

Unfortunately, it does not work for me with those versions, it generates this error:

```
Error: RollupError: node_modules/@storybook-vue/nuxt/node_modules/nuxt/dist./nuxt.js (4:9): "appId" is not exported by "virtual:nuxt:./.nuxt-storybook/nuxt.config.mjs", imported by "node_modules/@storybook-vue/nuxt/node_modules/nuxt/dist./nuxt.js".
file: ./node_modules/@storybook-vue/nuxt/node_modules/nuxt/dist./nuxt.js:4:9
```

**Comment by Yakov-Chernogor** [comment]
I'm also getting this error with those versions
```
ERROR  x Build failed in 3.80s

=> Failed to build the preview
Error: RollupError: node_modules/@storybook-vue/nuxt/node_modules/nuxt/dist/app/nuxt.js (4:9): "appId" is not exported by "virtual:nuxt:./.nuxt-storybook/nuxt.config.mjs", imported by "node_modules/@storybook-vue/nuxt/node_modules/nuxt/dist/app/nuxt.js".
file: ./node_modules/@storybook-vue/nuxt/node_modules/nuxt/dist/app/nuxt.js:4:9

2: import { createHooks } from "hookable";
3: import { getContext } from "unctx";
4: import { appId } from "#build/nuxt.config.mjs";
            ^
5: function getNuxtAppCtx(appName = appId || "nuxt-app") {
6:   return getContext(appName, {

    at defineNuxtConfig (./node_modules/@storybook-vue/nuxt/dist/preset.cjs:6752:11)
    at async Object.viteFinal (./node_modules/@storybook-vue/nuxt/dist/preset.cjs:6773:22)
    at async createViteServer (./node_modules/@storybook/builder-vite/dist/index.js:55:11761)
    at async Module.start (./node_modules/@storybook/builder-vite/dist/index.js:58:988)

WARN Broken build, fix the error above.
WARN You may need to refresh the browser.

error Command failed with exit code 1.
```

**Comment by andresilva-cc** [comment]
@Yakov-Chernogor Ok, I fixed that. Just add to your `package.json` and reinstall the dependencies:

```
  "resolutions": {
    "nuxt": "3.11.2"
  }
```

That's because `@storybook-vue/nuxt` installs version 3.12 of Nuxt, so adding that to the `package.json` makes all dependencies stick to Nuxt 3.11.2.

**Comment by Limonische** [comment]
Thanks everyone, will test it out

**Comment by mehcode** [comment]
@tobiasdiez I played with your PR for awhile. If I take it as-is and then patch the `preset.mjs` file to the following snippet - everything _seems_ to work. This is definitely a hack. But I hope showing something that works will help.

```ts
import { setup } from '@storybook/vue3';
import { reactive } from "vue";

const appConfig = reactive({ baseURL: "/" });

window.__NUXT__ = {
  serverRendered: false,
  config: {
    public: {},
    app: appConfig,
  },
  data: {},
  state: {},
};

setup(async (vueApp) => {
  const { createNuxtApp, applyPlugins } = await import("nuxt/app");
  const { getContext } = await import("unctx");
  const { $fetch } = await import("ofetch");
  const { runtimeConfig } = await import("virtual:nuxt-runtime-config");

  window.__NUXT__ = {
    serverRendered: false,
    config: {
      public: {},
      app: appConfig,
      ...runtimeConfig,
    },
    data: {},
    state: {},
  };

  const nuxtAppName = `nuxt-app`;
  const nuxtCtx = getContext(nuxtAppName);
  if (nuxtCtx.tryUse()) {
    return;
  }

  if (!globalThis.$fetch) {
    globalThis.$fetch = $fetch.create({
      baseURL: "/"
    });
  }

  const nuxt = createNuxtApp({
    vueApp,
    globalName: nuxtAppName
  });

  await nuxt.hooks.callHook("app:created", vueApp);
  await nuxt.hooks.callHook("app:beforeMount", vueApp);
  nuxtCtx.set(nuxt, true);

  await applyPlugins(nuxt, (await import("#build/plugins")).default);
  await import("#build/css");

  Object.assign(appConfig, window.__NUXT__.config.app);
});
```

FYI if anyone wants to try this, I used `pnpm patch @storybook-vue/nuxt` to apply this (and the PR updates by `pnpm build`-ing them and copying over the changes into a local patch.

**Comment by tobiasdiez** [comment]
Thanks everyone for testing!

@cheindl @obulat The `cdnURL` error seems unrelated, and I'll have a look at it later.

@mehcode why did you needed to apply these patches on top of it? What was not working without them?

**Comment by tobiasdiez** [comment]
Not sure if this is really fixed... I tried it again in a new project and it seems the error still occurs. Strange...

**Comment by mehcode** [comment]
> why did you needed to apply these patches on top of it? What was not working without them?

I had the same issues others have reported, the `Cannot read properties of undefined (reading 'cdnURL')` error.

From my testing/hacking I noticed that `window.__NUXT__` was being read before you're setting it and the value of `.config.app` is being stored.

**Comment by mehcode** [comment]
I should also mention a couple other things. 

- I'm not sure if it was these changes you've made to fix this, or Nuxt 3.12 in general, but the `viteFinal` from `@storybook/vue3-vite` is not being merged properly. I had to manually add `vue: isStorybook ? "vue/dist/vue.esm-bundler.js" : "vue",` to `alias` in my nuxt configuration.

- Additionally, this only fixes the dev storybook, the static version required me to add the following to my `preview` file. It's some kind of data race. Without this, I was getting an error "no story set".

```
// HACK: ever since the update to Nuxt 3.12 and Storybook 8.12, the static
//  build of our storybook fails to select the story on first load
//  this hacks into the internal API of storybook to force a select
//  immediately
setTimeout(() => {
  const query = new URLSearchParams(window.parent.location.search.slice(1));
  const storyId = query.get("path")?.split("/")?.at(-1);

  if (storyId != null) {
    addons.getChannel().emit("setCurrentStory", {
      storyId,
    });
  }
}, 0);
```

**Comment by mehcode** [comment]
So the first issue definitely looks like an issue with this package: https://github.com/nuxt-modules/storybook/blob/0a78a2770652c3d681c6d7ad9d75e52d21118326/packages/storybook-addon/src/preset.ts#L321-L331

We're ignoring the configuration entirely here.

**Comment by mehcode** [comment]
Here - we are using it correctly - https://github.com/nuxt-modules/storybook/blob/v8.1.0/packages/storybook-nuxt/src/preset.ts#L111

**Comment by mehcode** [comment]
If you'll give me an hour, I'll send a patch to fix part of this. The logic around config merging is broken on `main` and that's what I'm looking at right now.

**Comment by mehcode** [comment]
@tobiasdiez This fixes the first remaining issue I had - https://github.com/nuxt-modules/storybook/pull/740

I'm still using the hacked preview file locally though

**Comment by tobiasdiez** [comment]
This issue has been resolved in the latest nightly release. For installation instructions, please refer to: https://storybook.nuxtjs.org/getting-started/setup#installation. If you experience any further problems, don't hesitate to reopen this or create a new issue.

**Comment by Nemhis** [comment]
Still can not. Step by step installed by [nightly build instructions](https://storybook.nuxtjs.org/getting-started/setup#nightly-builds)

Error:
```
✘ [ERROR] Could not resolve "virtual:nuxt-runtime-config"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30:
      5 │ import { runtimeConfig } from 'virtual:nuxt-runtime-config';
        ╵                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  You can mark the path "virtual:nuxt-runtime-config" as external to exclude it from the bundle,
  which will remove this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/css"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7:
      6 │ import '#build/css';
        ╵        ~~~~~~~~~~~~

  You can mark the path "#build/css" as external to exclude it from the bundle, which will remove
  this error and leave the unresolved path in the bundle.

✘ [ERROR] Could not resolve "#build/plugins"

    node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20:
      7 │ import plugins from '#build/plugins';
        ╵                     ~~~~~~~~~~~~~~~~

  You can mark the path "#build/plugins" as external to exclude it from the bundle, which will
  remove this error and leave the unresolved path in the bundle.


 ERROR  error while updating dependencies:                                                                                                                                                  3:57:11 PM
Error: Build failed with 3 errors:
node_modules/@storybook-vue/nuxt/dist/preview.mjs:5:30: ERROR: Could not resolve "virtual:nuxt-runtime-config"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:6:7: ERROR: Could not resolve "#build/css"
node_modules/@storybook-vue/nuxt/dist/preview.mjs:7:20: ERROR: Could not resolve "#build/plugins"
```

Package.json
```json
...
"dependencies": {
    "@healerlab/vue3-facebook-login": "^0.1.7",
    "@nuxt/eslint": "^0.3.13",
    "@nuxtjs/i18n": "^8.3.1",
    "@pinia/nuxt": "^0.5.1",
    "@vee-validate/i18n": "^4.13.1",
    "@vee-validate/nuxt": "^4.13.1",
    "@vee-validate/rules": "^4.13.1",
    "@vkid/sdk": "^2.1.0",
    "@vueuse/core": "^10.11.0",
    "lodash-es": "^4.17.21",
    "nuxt": "^3.12.1",
    "nuxt-viewport": "^2.1.5",
    "nuxt-vue3-google-signin": "^0.0.11",
    "pinia": "^2.1.7",
    "vue": "^3.4.27",
    "vue-router": "^4.3.3",
    "vue3-recaptcha-v2": "2.0.2"
  },
  "devDependencies": {
    "@nuxtjs/storybook": "npm:@nuxtjs/storybook@nightly",
    "@nuxtjs/svg-sprite": "1.0.2",
    "@storybook-vue/nuxt": "0.2.10",
    "@storybook/addon-essentials": "8.2.8",
    "@storybook/addon-interactions": "8.2.8",
    "@storybook/addon-links": "8.2.8",
    "@storybook/blocks": "8.2.8",
    "@storybook/builder-vite": "8.2.8",
    "@storybook/test": "8.2.8",
    "@storybook/vue3": "8.2.8",
    "@storybook/vue3-vite": "8.2.8",
    "@types/lodash-es": "4.17.12",
    "@types/node": "^18.17.5",
    "@typescript-eslint/eslint-plugin": "^7.13.1",
    "@typescript-eslint/parser": "^7.13.1",
    "eslint": "^8.57.0",
    "sass": "^1.77.5",
    "sass-loader": "^14.2.1",
    "storybook": "8.2.8",
    "vite-svg-loader": "5.1.0"
  }
...
```

**Comment by tobiasdiez** [comment]
the dev deps `@storybook-vue/nuxt @storybook/builder-vite @storybook/vue3 @storybook/vue3-vite` are not needed anymore. can you try after removing them

**Comment by Nemhis** [comment]
Whaty should i use as framework in `.storybook/main.ts` ? Now it look like this:
```ts
const config = {
  stories: [
    '../stories/**/*.mdx',
    '../stories/**/*.stories.@(js|jsx|mjs|ts|tsx)',
  ],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/addon-interactions',
  ],
  framework: {
    name: '@storybook-vue/nuxt',
    options: {},
  },
  docs: {
    autodocs: 'tag',
  },
}
export default config
```

**Comment by tobiasdiez** [comment]
Seams to be correct: https://github.com/nuxt-modules/storybook/tree/main/examples/starter

**Comment by Nemhis** [comment]
It seems to be working. Thank you so much.

**Comment by tobiasdiez** [comment]
Since this seems to work well, could you please release a new version @chakAs3

**Comment by AlexanderBredun** [comment]
Yeah, even in a minimal reproduction, it gives that error for any image in the public directory. It seems to be related to this part, which I guess you have been changing in this thread.
![image](https://github.com/user-attachments/assets/0d32bd46-1b9c-4449-9cba-894d42bf620d)

**Comment by tobiasdiez** [comment]
Do you have any visible problems from this error, or is it only annoying in the console? Could you please open a new issue, since it's not directly related to this one here.

My first guess is that the setup functions are run in parallel and thus the nuxt config is not yet initialized. (see https://github.com/nuxt-modules/storybook/issues/642#issuecomment-2195482436)

**Comment by AlexanderBredun** [comment]
i created new issue https://github.com/nuxt-modules/storybook/issues/753 and no with this error storie doesnt work at all

**Comment by tkjaergaard** [comment]
After using the nightly build (^8.2.0-28726259.a3fa103) i get this warning in the terminal:

```bash
WARN The following packages are incompatible with Storybook 8.2.7 as they depend on different major versions of Storybook packages:
WARN - @nuxtjs/storybook@8.2.0-28726259.a3fa103
WARN  Repo: https://storybook.nuxtjs.org
WARN - @storybook/addon-essentials@8.2.9
WARN  Repo: https://github.com/storybookjs/storybook/tree/next/code/addons/essentials
WARN - @storybook/addon-links@8.2.9
WARN  Repo: https://github.com/storybookjs/storybook/tree/next/code/addons/links
WARN - @storybook/types@8.2.9
WARN  Repo: https://github.com/storybookjs/storybook/tree/next/code/lib/types
WARN 
WARN 
WARN Please consider updating your packages or contacting the maintainers for compatibility details.
WARN For more on Storybook 8 compatibility, see the linked GitHub issue:
WARN https://github.com/storybookjs/storybook/issues/26031
```

And still getting the error "Failed to fetch dynamically imported module: http://localhost:6006/node_modules/.cache/storybook/1c3385a5d25e538d10b518b310c74d3ca2690b6aaffeadccd74da79736171f86/sb-vite/deps/@storybook_addon-essentials_outline_preview.js?v=d7058fda"

Using `nuxt 3.12.4`

**Comment by obulat** [comment]
@tobiasdiez , it seems that the paths problems arise when the Nuxt app has `/pages`. If you run `/examples/showcase`, you can see that there are errors stating that the route cannot be found. It seems that the routes from pages are overwriting the routes created by the Storybook Nuxt app.
